### PR TITLE
FEATURE: Reward Fees

### PR DESCRIPTION
**Pull request by coachchucksol**
To address Issue #46 `reward_fees_bps` were added closely following the design of the spl-stake-pool's `epoch_fee` design.

### CODE REVIEW

File: vault_program/src/update_vault_balance.rs [hunk 1]
```diff
@@ -1,28 +1,78 @@
...
+                )?,
+                &[
+                    vrt_mint.clone(),
+                    vault_fee_token_account.clone(),
+                    vault_info.clone(),
+                ],
+                &[&seed_slices],
+            )?;
+        }
+    }
```
Comment (buffalu) [inline vault_program/src/update_vault_balance.rs:66]: need to increment the vrt_supply as well

Comment (buffalu) [reply vault_program/src/update_vault_balance.rs:66]: can you check for that in tests too?

File: vault_program/src/update_vault_balance.rs [hunk 2]
```diff
@@ -1,28 +1,78 @@
...
+
+        load_token_mint(vrt_mint)?;
+        load_associated_token_account(vault_token_account, vault_info.key, &vault.supported_mint)?;
+        load_associated_token_account(vault_fee_token_account, &vault.fee_wallet, vrt_mint.key)?;
+        load_token_program(token_program)?;
+        vault.check_vrt_mint(vrt_mint.key)?;
+
+        // Calculate rewards
+
+        let reward_fee = vault.calculate_rewards_fee(new_balance)?;
```
Comment (buffalu) [inline vault_program/src/update_vault_balance.rs:38]: i think this reward fee needs to be in terms of the VRT, not the supported mint?

File: vault_core/src/vault.rs [hunk 3]
```diff
@@ -342,6 +347,23 @@ impl Vault {
     // Minting and burning
     // ------------------------------------------
 
+    /// Calculate the rewards fee. This is used in `update_vault_balance` to mint
+    /// the `fee` amount in VRTs to the `fee_wallet`.
+    pub fn calculate_rewards_fee(&self, new_balance: u64) -> Result<u64, VaultError> {
```
Comment (buffalu) [inline vault_core/src/vault.rs:352]: look at stake pool calc_epoch_fee_amount, the returned number needs to be in terms of VRT

File: integration_tests/tests/fixtures/vault_client.rs [hunk 4]
```diff
@@ -1038,12 +1047,18 @@ impl VaultProgramClient {
...
+                &get_associated_token_address(&vault.fee_wallet, &vault.vrt_mint),
+                &spl_token::ID,
             )],
             Some(&self.payer.pubkey()),
             &[&self.payer],
             blockhash,
         ))
         .await?;
+
+        println!("vault balance updated");
```
[the remark this answers is not in the document]
Comment (coachchucksol) [reply integration_tests/tests/fixtures/vault_client.rs:1060]: Done

File: integration_tests/tests/fixtures/vault_client.rs [hunk 5]
```diff
@@ -1570,6 +1585,69 @@ impl VaultProgramClient {
...
+            .banks_client
+            .get_account(vault_fee_token_account)
+            .await
+            .unwrap()
+            .unwrap();
+
+        Ok(SPLTokenAccount::unpack(&account.data).unwrap())
+    }
+
+    pub async fn reward_vault(
```
[the remark this answers is not in the document]
Comment (coachchucksol) [reply integration_tests/tests/fixtures/vault_client.rs:1608]: Done

File: integration_tests/tests/vault/reward_fee.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            .reward_vault(&vault_root.vault_pubkey, &rewarder, MINT_AMOUNT)
+            .await
+            .unwrap();
+
+        let config = vault_program_client
+            .get_config(&Config::find_program_address(&jito_vault_program::id()).0)
+            .await
+            .unwrap();
+        fixture
+            .warp_slot_incremental(2 * config.epoch_length)
```
[the remark this answers is not in the document]
Comment (coachchucksol) [reply integration_tests/tests/vault/reward_fee.rs:53]: Done

File: integration_tests/tests/vault/reward_fee.rs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            .unwrap();
+
+        let reward_fee_account = vault_program_client
+            .get_reward_fee_token_account(&vault_root.vault_pubkey)
+            .await
+            .unwrap();
+
+        assert_eq!(MINT_AMOUNT, vault.tokens_deposited);
+        assert_eq!(0, reward_fee_account.amount);
+    }
```
[the remark this answers is not in the document]
Comment (coachchucksol) [reply integration_tests/tests/vault/reward_fee.rs:213]: The balance increases, but not by `update_vault_balance`, when a depositor deposits they call `mint_to` which in `mint_with_fee` updates `tokens_deposited`

File: vault_program/src/update_vault_balance.rs (hunk 1 above)
Comment (coachchucksol) [reply vault_program/src/update_vault_balance.rs:66]: Done

File: vault_program/src/update_vault_balance.rs (hunk 2 above)
Comment (coachchucksol) [reply vault_program/src/update_vault_balance.rs:38]: Done

File: vault_core/src/vault.rs (hunk 3 above)
Comment (coachchucksol) [reply vault_core/src/vault.rs:352]: Done

Comment (coachchucksol) [reply vault_core/src/vault.rs:352]:
```

    pub fn calculate_rewards_fee(&self, new_balance: u64) -> Result<u64, VaultError> {
        let rewards = new_balance.saturating_sub(self.tokens_deposited);

        let vrt_rewards = self.calculate_vrt_mint_amount(rewards)?;

        if vrt_rewards == 0 {
            return Ok(0);
        }

        let fee = vrt_rewards
            .checked_mul(self.reward_fee_bps as u64)
            .ok_or(VaultError::VaultOverflow)?
            .div_ceil(10_000);

        Ok(fee)
    }
```

Comment (coachchucksol) [reply vault_core/src/vault.rs:352]: Is this sufficient?

File: integration_tests/tests/vault/reward_fee.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (coachchucksol) [reply integration_tests/tests/vault/reward_fee.rs:309]: Done

File: integration_tests/tests/vault/reward_fee.rs [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+            .get_vault(&vault_root.vault_pubkey)
+            .await
+            .unwrap();
+
+        let reward_fee_account = vault_program_client
+            .get_reward_fee_token_account(&vault_root.vault_pubkey)
+            .await
+            .unwrap();
+
+        println!("TOKENS DEPOSITED {}", vault.tokens_deposited);
```
Comment (buffalu) [inline integration_tests/tests/vault/reward_fee.rs:303]: rm

File: vault_program/src/update_vault_balance.rs [hunk 9]
```diff
@@ -1,28 +1,80 @@
...
 
     Config::load(program_id, config, false)?;
     Vault::load(program_id, vault_info, true)?;
-    let mut vault_data = vault_info.data.borrow_mut();
-    let vault = Vault::try_from_slice_unchecked_mut(&mut vault_data)?;
-    load_associated_token_account(vault_token_account, vault_info.key, &vault.supported_mint)?;
 
-    // TODO (LB): pay out fee account for any accrued fees to vault fee wallet
-    vault.tokens_deposited = Account::unpack(&vault_token_account.data.borrow())?.amount;
+    let new_balance = Account::unpack(&vault_token_account.data.borrow())?.amount;
```
Comment (buffalu) [inline vault_program/src/update_vault_balance.rs:23]: nit: loading token account before parsing it

File: vault_program/src/update_vault_balance.rs [hunk 10]
```diff
@@ -1,28 +1,80 @@
...
+    };
+
+    // Update state
+    {
+        // need to drop the reference to vault_data before we can borrow_mut again
+        let mut vault_data = vault_info.data.borrow_mut();
+        let vault = Vault::try_from_slice_unchecked_mut(&mut vault_data)?;
+
+        vault.tokens_deposited = new_balance;
+        vault.vrt_supply = vault.vrt_supply.checked_add(reward_fee).unwrap();
```
Comment (buffalu) [inline vault_program/src/update_vault_balance.rs:76]: ok_or

File: integration_tests/tests/vault/reward_fee.rs (hunk 8 above)
Comment (coachchucksol) [reply integration_tests/tests/vault/reward_fee.rs:303]: Done

File: vault_program/src/update_vault_balance.rs (hunk 9 above)
Comment (coachchucksol) [reply vault_program/src/update_vault_balance.rs:23]: Done

File: vault_program/src/update_vault_balance.rs (hunk 10 above)
Comment (coachchucksol) [reply vault_program/src/update_vault_balance.rs:76]: Done

**Pull request merged**